File: compiler/bench/pollardrho.fs
\ expected: 6020697
\ Pollard's rho - find factors using cycle detection

: gcd ( a b -- gcd )
  begin dup while tuck mod repeat drop ;

: abs ( n -- |n| )
  dup 0< if negate then ;

: pollard-f ( x n -- x' )
  swap dup * 1+ swap mod ;

: pollard-rho ( n -- factor )
  dup 1 and 0= if drop 2 exit then
  2 2 rot                 \ x=2, y=2, n
  begin
    over 2 pick pollard-f rot drop swap  \ x'=f(x), y, n
    over 2 pick pollard-f 2 pick pollard-f
    rot drop swap         \ x, y'=f(f(y)), n
    2dup - abs 2 pick gcd \ x y n d
    dup 1 = while drop
  repeat
  nip nip nip ;

: main
  \ Sum smallest factors of odd numbers
  0
  10000 3 do
    i pollard-rho +
  2 +loop
  . cr ;

File: compiler/bench/pollardrho.c
// expected: 6020697
// Pollard's rho - find factors using cycle detection
#include <stdio.h>
#include <stdlib.h>

long gcd(long a, long b) {
    while (b) { long t = b; b = a % b; a = t; }
    return a;
}

long pollard_f(long x, long n) {
    return (x * x + 1) % n;
}

long pollard_rho(long n) {
    if ((n & 1) == 0) return 2;
    long x = 2, y = 2;
    long d = 1;
    while (d == 1) {
        x = pollard_f(x, n);
        y = pollard_f(pollard_f(y, n), n);
        d = gcd(labs(x - y), n);
    }
    return d;
}

int main(void) {
    long sum = 0;
    for (long i = 3; i < 10000; i += 2) {
        sum += pollard_rho(i);
    }
    printf("%ld\n", sum);
    return 0;
}
